Error subscribing to @OnClose handler for WebSocket bean [..Controller]  Broadcast Failure: null

### Expected Behavior

_No response_

### Actual Behaviour

When in a browser more then 4/5 tabs are opened and connected with socket. On Closing browser this following exception is thrown.

```
09:11:23.136 [globalEventExecutor-2-6488] ERROR i.m.h.s.n.w.NettyServerWebSocketHandler - Error subscribing to @OnClose handler for WebSocket bean [notification.hub.web
socket.controller.WebsocketController@4c5f9154]: Broadcast Failure: null
io.micronaut.websocket.exceptions.WebSocketSessionException: Broadcast Failure: null
        at io.micronaut.http.netty.websocket.NettyServerWebSocketBroadcaster.lambda$null$2(NettyServerWebSocketBroadcaster.java:84)
        at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:578)
        at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:552)
        at io.netty.util.concurrent.DefaultPromise.access$200(DefaultPromise.java:35)
        at io.netty.util.concurrent.DefaultPromise$1.run(DefaultPromise.java:502)
        at io.netty.util.concurrent.GlobalEventExecutor$TaskRunner.run(GlobalEventExecutor.java:240)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:831)
Caused by: io.netty.channel.group.ChannelGroupException: null
        at io.netty.channel.group.DefaultChannelGroupFuture$1.operationComplete(DefaultChannelGroupFuture.java:73)
        at io.netty.channel.group.DefaultChannelGroupFuture$1.operationComplete(DefaultChannelGroupFuture.java:48)
        at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:578)
        at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:552)
        at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:491)
        at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:616)
        at io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:609)
        at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:117)
        at io.netty.handler.stream.ChunkedWriteHandler$PendingWrite.fail(ChunkedWriteHandler.java:366)
        at io.netty.handler.stream.ChunkedWriteHandler.discard(ChunkedWriteHandler.java:186)
        at io.netty.handler.stream.ChunkedWriteHandler.doFlush(ChunkedWriteHandler.java:194)
        at io.netty.handler.stream.ChunkedWriteHandler.flush(ChunkedWriteHandler.java:131)
        at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:750)
        at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:742)
        at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:728)
        at io.netty.channel.ChannelOutboundHandlerAdapter.flush(ChannelOutboundHandlerAdapter.java:125)
        at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:750)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:765)
        at io.netty.channel.AbstractChannelHandlerContext$WriteTask.run(AbstractChannelHandlerContext.java:1071)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
        ... 3 common frames omitted
 __  __ _

```

### Steps To Reproduce

Create some sockets in Browser and then close the browser tab. Example
```
## in Chrome browser
for(i=0;i<3;i++){
    new WebSocket("ws://localhost:8080/socket/extra");
}

```


### Example Application
This exception only occur if we have 'close' annotation method.

https://github.com/micronaut-projects/micronaut-examples/blob/master/websocket-chat/src/main/java/websocket/chat/ChatWebSocket.java

### Version

2.5.12 Same problem in 3.0.3

--- a/http-netty/src/main/java/io/micronaut/http/netty/websocket/NettyServerWebSocketBroadcaster.java
+++ b/http-netty/src/main/java/io/micronaut/http/netty/websocket/NettyServerWebSocketBroadcaster.java
@@ -16,16 +16,22 @@
 package io.micronaut.http.netty.websocket;
 
 import io.micronaut.context.annotation.Requires;
+import io.micronaut.core.annotation.Nullable;
 import io.micronaut.http.MediaType;
 import io.micronaut.websocket.WebSocketBroadcaster;
 import io.micronaut.websocket.WebSocketSession;
 import io.micronaut.websocket.exceptions.WebSocketSessionException;
+import io.netty.channel.Channel;
+import io.netty.channel.group.ChannelGroupException;
 import io.netty.handler.codec.http.websocketx.WebSocketFrame;
 import io.netty.util.Attribute;
 import jakarta.inject.Singleton;
 import org.reactivestreams.Publisher;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.FluxSink;
+
+import java.nio.channels.ClosedChannelException;
+import java.util.Map;
 import java.util.function.Predicate;
 
 /**
@@ -76,17 +82,53 @@ public class NettyServerWebSocketBroadcaster implements WebSocketBroadcaster {
                     NettyWebSocketSession s = attr.get();
                     return s != null && s.isOpen() && filter.test(s);
                 }).addListener(future -> {
-                    if (future.isSuccess()) {
-                        emitter.next(message);
-                        emitter.complete();
-                    } else {
-                        Throwable cause = future.cause();
-                        emitter.error(new WebSocketSessionException("Broadcast Failure: " + cause.getMessage(), cause));
+                    if (!future.isSuccess()) {
+                        Throwable cause = extractBroadcastFailure(future.cause());
+                        if (cause != null) {
+                            emitter.error(new WebSocketSessionException("Broadcast Failure: " + cause.getMessage(), cause));
+                            return;
+                        }
                     }
+                    emitter.next(message);
+                    emitter.complete();
                 });
             } catch (Throwable e) {
                 emitter.error(new WebSocketSessionException("Broadcast Failure: " + e.getMessage(), e));
             }
         }, FluxSink.OverflowStrategy.BUFFER);
+    }
+
+    /**
+     * Attempt to extract a single failure from a failure of {@link io.netty.channel.group.ChannelGroup#write}
+     * exception. {@link io.netty.channel.group.ChannelGroup} aggregates exceptions into a {@link ChannelGroupException}
+     * that has no useful stacktrace. If there was only one actual failure, we will just forward that instead of the
+     * {@link ChannelGroupException}.
+     *
+     * We also need to ignore {@link ClosedChannelException}s.
+     */
+    @Nullable
+    private Throwable extractBroadcastFailure(Throwable failure) {
+        if (failure instanceof ChannelGroupException) {
+            Throwable singleCause = null;
+            for (Map.Entry<Channel, Throwable> entry : (ChannelGroupException) failure) {
+                Throwable entryCause = extractBroadcastFailure(entry.getValue());
+                if (entryCause != null) {
+                    if (singleCause == null) {
+                        singleCause = entryCause;
+                    } else {
+                        return failure;
+                    }
+                }
+            }
+            return singleCause;
+        } else if (failure instanceof ClosedChannelException) {
+            // ClosedChannelException can happen when there is a race condition between the call to writeAndFlush and
+            // the closing of a channel. session.isOpen will still return true, but when the write is actually
+            // performed, the channel is closed. Since we would have skipped the write anyway had we known the channel
+            // would go away, we can safely ignore this error.
+            return null;
+        } else {
+            return failure;
+        }
     }
 }

--- a/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/websocket/BroadcasterSpec.groovy
+++ b/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/websocket/BroadcasterSpec.groovy
@@ -123,13 +123,9 @@ class BroadcasterSpec extends Specification {
         @OnMessage
         def onMessage(String msg) {}
 
-        @OnOpen
-        def onOpen() {
-            connectionPhaser.register()
-        }
-
         @OnClose
         def onClose() {
+            connectionPhaser.register()
             broadcaster.broadcast('foo').subscribe(new Subscriber<String>() {
                 @Override
                 void onSubscribe(Subscription s) {

--- a/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/websocket/BroadcasterSpec.groovy
+++ b/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/websocket/BroadcasterSpec.groovy
@@ -34,6 +34,7 @@ import spock.lang.Issue
 import spock.lang.Specification
 
 import java.util.concurrent.Phaser
+import java.util.concurrent.TimeUnit
 
 class BroadcasterSpec extends Specification {
     @Issue('https://github.com/micronaut-projects/micronaut-core/issues/6269')
@@ -101,7 +102,7 @@ class BroadcasterSpec extends Specification {
         }
 
         def server = ctx.getBean(Server)
-        server.connectionPhaser.arriveAndAwaitAdvance()
+        server.connectionPhaser.awaitAdvanceInterruptibly(server.connectionPhaser.arrive(), 10, TimeUnit.SECONDS)
 
         expect:
         server.errors.isEmpty()

--- a/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/websocket/BroadcasterSpec.groovy
+++ b/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/websocket/BroadcasterSpec.groovy
@@ -1,0 +1,156 @@
+package io.micronaut.http.server.netty.websocket
+
+import io.micronaut.context.ApplicationContext
+import io.micronaut.context.annotation.Requires
+import io.micronaut.runtime.server.EmbeddedServer
+import io.micronaut.websocket.WebSocketBroadcaster
+import io.micronaut.websocket.annotation.OnClose
+import io.micronaut.websocket.annotation.OnMessage
+import io.micronaut.websocket.annotation.OnOpen
+import io.micronaut.websocket.annotation.ServerWebSocket
+import io.netty.bootstrap.Bootstrap
+import io.netty.channel.Channel
+import io.netty.channel.ChannelFuture
+import io.netty.channel.ChannelHandlerContext
+import io.netty.channel.ChannelInitializer
+import io.netty.channel.SimpleChannelInboundHandler
+import io.netty.channel.nio.NioEventLoopGroup
+import io.netty.channel.socket.SocketChannel
+import io.netty.channel.socket.nio.NioSocketChannel
+import io.netty.handler.codec.http.DefaultHttpHeaders
+import io.netty.handler.codec.http.FullHttpResponse
+import io.netty.handler.codec.http.HttpClientCodec
+import io.netty.handler.codec.http.HttpObjectAggregator
+import io.netty.handler.codec.http.websocketx.CloseWebSocketFrame
+import io.netty.handler.codec.http.websocketx.WebSocketClientHandshakerFactory
+import io.netty.handler.codec.http.websocketx.WebSocketCloseStatus
+import io.netty.handler.codec.http.websocketx.WebSocketVersion
+import jakarta.inject.Inject
+import jakarta.inject.Singleton
+import org.jetbrains.annotations.NotNull
+import org.reactivestreams.Subscriber
+import org.reactivestreams.Subscription
+import spock.lang.Issue
+import spock.lang.Specification
+
+import java.util.concurrent.Phaser
+
+class BroadcasterSpec extends Specification {
+    @Issue('https://github.com/micronaut-projects/micronaut-core/issues/6269')
+    def 'closing channels simultaneously does not log an error'() {
+        given:
+        def ctx = ApplicationContext.run([
+                'spec.name'                            : 'BroadcasterSpec',
+                'micronaut.server.netty.worker.threads': 1,
+                'micronaut.http.client.num-of-threads' : 1,
+                'micronaut.server.port'                : -1
+        ])
+        def embeddedServer = ctx.getBean(EmbeddedServer)
+        embeddedServer.start()
+
+        def uri = "ws://$embeddedServer.host:$embeddedServer.port/ws"
+
+        def channels = new ArrayList<Channel>()
+        def handshakes = new ArrayList<ChannelFuture>()
+        def eventLoopGroup = new NioEventLoopGroup(1)
+        Bootstrap bootstrap = new Bootstrap()
+                .group(eventLoopGroup)
+                .channel(NioSocketChannel)
+                .handler(new ChannelInitializer<SocketChannel>() {
+                    @Override
+                    protected void initChannel(@NotNull SocketChannel ch) throws Exception {
+                        def handshaker = WebSocketClientHandshakerFactory.newHandshaker(new URI(uri), WebSocketVersion.V13, null, true, new DefaultHttpHeaders())
+                        def handshakeFuture = ch.newPromise()
+                        handshakes.add(handshakeFuture)
+                        ch.pipeline()
+                            .addLast(new HttpClientCodec())
+                            .addLast(new HttpObjectAggregator(1024))
+                            .addLast(new SimpleChannelInboundHandler<Object>() {
+                                @Override
+                                protected void channelRead0(ChannelHandlerContext ctx_, Object msg) throws Exception {
+                                    if (!handshaker.isHandshakeComplete()) {
+                                        handshaker.finishHandshake(ctx_.channel(), (FullHttpResponse) msg)
+                                        handshakeFuture.setSuccess()
+                                    }
+                                }
+
+                                @Override
+                                void channelActive(@NotNull ChannelHandlerContext ctx_) throws Exception {
+                                    handshaker.handshake(ctx_.channel())
+                                }
+                            })
+                        channels.add(ch)
+                    }
+                })
+                .remoteAddress(embeddedServer.host, embeddedServer.port)
+
+        for (int i = 0; i < 3; i++) {
+            bootstrap.connect().sync()
+        }
+        for (def handshake : handshakes) {
+            handshake.sync()
+        }
+        eventLoopGroup.execute {
+            for (Channel channel : channels) {
+                channel.writeAndFlush(new CloseWebSocketFrame(WebSocketCloseStatus.ENDPOINT_UNAVAILABLE))
+                channel.close()
+            }
+        }
+        for (Channel channel : channels) {
+            channel.closeFuture().sync()
+        }
+
+        def server = ctx.getBean(Server)
+        server.connectionPhaser.arriveAndAwaitAdvance()
+
+        expect:
+        server.errors.isEmpty()
+
+        cleanup:
+        ctx.close()
+    }
+
+    @Singleton
+    @Requires(property = 'spec.name', value = 'BroadcasterSpec')
+    @ServerWebSocket("/ws")
+    static class Server {
+        @Inject
+        WebSocketBroadcaster broadcaster
+        List<Throwable> errors = Collections.synchronizedList(new ArrayList<>())
+        Phaser connectionPhaser = new Phaser(1) // one party for waiting
+
+        @OnMessage
+        def onMessage(String msg) {}
+
+        @OnOpen
+        def onOpen() {
+            connectionPhaser.register()
+        }
+
+        @OnClose
+        def onClose() {
+            broadcaster.broadcast('foo').subscribe(new Subscriber<String>() {
+                @Override
+                void onSubscribe(Subscription s) {
+                    s.request(Long.MAX_VALUE)
+                }
+
+                @Override
+                void onNext(String s) {
+                }
+
+                @Override
+                void onError(Throwable t) {
+                    t.printStackTrace()
+                    errors.add(t)
+                    connectionPhaser.arrive()
+                }
+
+                @Override
+                void onComplete() {
+                    connectionPhaser.arrive()
+                }
+            })
+        }
+    }
+}

--- a/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/websocket/BroadcasterSpec.groovy
+++ b/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/websocket/BroadcasterSpec.groovy
@@ -102,7 +102,7 @@ class BroadcasterSpec extends Specification {
         }
 
         def server = ctx.getBean(Server)
-        server.connectionPhaser.awaitAdvanceInterruptibly(server.connectionPhaser.arrive(), 10, TimeUnit.SECONDS)
+        eventLoopGroup.shutdownGracefully().sync()
 
         expect:
         server.errors.isEmpty()
@@ -118,14 +118,12 @@ class BroadcasterSpec extends Specification {
         @Inject
         WebSocketBroadcaster broadcaster
         List<Throwable> errors = Collections.synchronizedList(new ArrayList<>())
-        Phaser connectionPhaser = new Phaser(1) // one party for waiting
 
         @OnMessage
         def onMessage(String msg) {}
 
         @OnClose
         def onClose() {
-            connectionPhaser.register()
             broadcaster.broadcast('foo').subscribe(new Subscriber<String>() {
                 @Override
                 void onSubscribe(Subscription s) {
@@ -140,12 +138,10 @@ class BroadcasterSpec extends Specification {
                 void onError(Throwable t) {
                     t.printStackTrace()
                     errors.add(t)
-                    connectionPhaser.arrive()
                 }
 
                 @Override
                 void onComplete() {
-                    connectionPhaser.arrive()
                 }
             })
         }

--- a/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/websocket/BroadcasterSpec.groovy
+++ b/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/websocket/BroadcasterSpec.groovy
@@ -2,6 +2,9 @@ package io.micronaut.http.server.netty.websocket
 
 import io.micronaut.context.ApplicationContext
 import io.micronaut.context.annotation.Requires
+import io.micronaut.context.event.BeanCreatedEvent
+import io.micronaut.context.event.BeanCreatedEventListener
+import io.micronaut.http.netty.channel.ChannelPipelineCustomizer
 import io.micronaut.runtime.server.EmbeddedServer
 import io.micronaut.websocket.WebSocketBroadcaster
 import io.micronaut.websocket.annotation.OnClose
@@ -14,6 +17,8 @@ import io.netty.channel.ChannelFuture
 import io.netty.channel.ChannelHandlerContext
 import io.netty.channel.ChannelInitializer
 import io.netty.channel.SimpleChannelInboundHandler
+import io.netty.channel.group.ChannelGroup
+import io.netty.channel.group.DefaultChannelGroup
 import io.netty.channel.nio.NioEventLoopGroup
 import io.netty.channel.socket.SocketChannel
 import io.netty.channel.socket.nio.NioSocketChannel
@@ -25,6 +30,7 @@ import io.netty.handler.codec.http.websocketx.CloseWebSocketFrame
 import io.netty.handler.codec.http.websocketx.WebSocketClientHandshakerFactory
 import io.netty.handler.codec.http.websocketx.WebSocketCloseStatus
 import io.netty.handler.codec.http.websocketx.WebSocketVersion
+import io.netty.util.concurrent.Promise
 import jakarta.inject.Inject
 import jakarta.inject.Singleton
 import org.jetbrains.annotations.NotNull
@@ -102,7 +108,7 @@ class BroadcasterSpec extends Specification {
         }
 
         def server = ctx.getBean(Server)
-        eventLoopGroup.shutdownGracefully().sync()
+        ctx.getBean(WaitForClients).sync()
 
         expect:
         server.errors.isEmpty()
@@ -144,6 +150,32 @@ class BroadcasterSpec extends Specification {
                 void onComplete() {
                 }
             })
+        }
+    }
+
+    @Singleton
+    @Requires(property = 'spec.name', value = 'BroadcasterSpec')
+    static class WaitForClients implements BeanCreatedEventListener<ChannelPipelineCustomizer> {
+        List<Channel> channels = Collections.synchronizedList(new ArrayList<>())
+
+        @Override
+        ChannelPipelineCustomizer onCreated(BeanCreatedEvent<ChannelPipelineCustomizer> event) {
+            if (event.bean.isServerChannel()) {
+                event.bean.doOnConnect {
+                    channels.add(it.channel())
+                    return it
+                }
+            }
+            return event.bean
+        }
+
+        def sync() {
+            channels.forEach {
+                it.closeFuture().sync()
+            }
+            channels.forEach {
+                it.eventLoop().shutdownGracefully().sync()
+            }
         }
     }
 }

